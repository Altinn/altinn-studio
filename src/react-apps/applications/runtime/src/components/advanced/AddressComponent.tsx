import axios, { AxiosResponse } from 'axios';
import classNames = require('classnames');
import update from 'immutability-helper';
import * as React from 'react';

import '../../styles/AddressComponent.css';
import '../../styles/shared.css';
import { renderValidationMessagesForComponent } from '../../utils/render';

export interface IAddressComponentProps {
  component: IFormAddressComponent;
  formData: { [id: string]: string };
  handleDataChange: (value: any, key: string) => void;
  getTextResource: (key: string) => string;
  isValid?: boolean;
  simplified: boolean;
  validationMessages?: IComponentValidations;
}

interface IAddressValidationErrors {
  zipCode?: string;
  houseNumber?: string;
}

export interface IAddressComponentState {
  address: string;
  zipCode: string;
  postPlace: string;
  careOf: string;
  houseNumber: string;
  validations: IAddressValidationErrors;
  mounted: boolean;
}

export enum AddressKeys {
  address = 'address',
  zipCode = 'zipCode',
  postPlace = 'postPlace',
  careOf = 'careOf',
  houseNumber = 'houseNumber',
}

export class AddressComponent extends React.Component<IAddressComponentProps, IAddressComponentState> {
  constructor(_props: IAddressComponentProps) {
    super(_props);
    const formData = this.props.formData ? this.props.formData : {};
    this.state = {
      address: formData.address || '',
      zipCode: formData.zipCode || '',
      postPlace: formData.postPlace || '',
      careOf: formData.careOf || '',
      houseNumber: formData.houseNumber || '',
      validations: {
        zipCode: null,
        houseNumber: null,
      },
      mounted: false,
    };
    if (this.state.zipCode) {
      this.fetchPostPlace(this.state.zipCode);
    }
  }

  public componentWillUnmount: () => void = () => {
    this.setState({
      mounted: false,
    });
  }

  public componentDidMount: () => void = () => {
    this.setState({
      mounted: true,
    });
  }

  public fetchPostPlace: (zipCode: string) => void = (zipCode: string) => {
    if (zipCode.match(new RegExp('^[0-9]{4}$'))) {
      axios.get('https://api.bring.com/shippingguide/api/postalCode.json', {
        params: {
          clientUrl: window.location.href,
          pnr: zipCode,
        },
      }).then((response: AxiosResponse) => {
        if (response.data.valid && this.state.mounted) {
          this.setState({
            postPlace: response.data.result,
            validations: {
              zipCode: null,
            },
          }, () => {
            this.onBlurField(AddressKeys.postPlace);
          });
        } else if (this.state.mounted) {
          this.setState({
            postPlace: '',
            validations: {
              zipCode: 'Postnummer er ikke gyldig',
            },
          }, () => {
            this.onBlurField(AddressKeys.postPlace);
          });
        }
      });
      this.setState({
        postPlace: 'Laster...',
      });
    }
  }

  public validate: () => IAddressValidationErrors = () => {
    const { zipCode, houseNumber } = this.state;
    const validationErrors: IAddressValidationErrors = {
      zipCode: null,
      houseNumber: null,
    };
    if (!zipCode.match(new RegExp('^[0-9]{4}$')) && zipCode !== '') {
      validationErrors.zipCode = 'Postnummer er ikke gyldig';
    } else {
      validationErrors.zipCode = null;
    }
    if (!houseNumber.match(new RegExp('^[a-z,A-Z]{1}[0-9]{4}$')) && houseNumber !== '') {
      validationErrors.houseNumber = 'Bolignummer er ikke gyldig';
    } else {
      validationErrors.houseNumber = null;
    }
    return validationErrors;
  }

  public updateField: (key: AddressKeys, event: any) => void = (key: AddressKeys, event: any): void => {
    const changedFieldValue: string = event.target.value;
    const changedKey: string = AddressKeys[key];
    this.setState((state: IAddressComponentState) => update(state, {
      [changedKey]: {
        $set: changedFieldValue,
      },
    }));
  }

  public onBlurField: (key: AddressKeys) => void = (key: AddressKeys) => {
    const validationErrors: IAddressValidationErrors = this.validate();
    if (!validationErrors.zipCode && !validationErrors.houseNumber) {
      this.props.handleDataChange(this.state[key], key);
      if (AddressKeys[key] === 'zipCode') {
        this.fetchPostPlace(this.state[key]);
      }
    }
    this.setState({
      validations: validationErrors,
    });
  }

  public joinValidationMessages = (): IComponentValidations => {
    const { validations } = this.state as any;
    let { validationMessages } = this.props;

    for (const fieldKey in validations) {
      if (!validations[fieldKey]) {
        continue;
      }

      if (!validationMessages) {
        validationMessages = {
          [fieldKey]: {
            errors: [],
            warnings: [],
          },
        };
      }

      if (validationMessages[fieldKey]) {
        validationMessages[fieldKey].errors.push(validations[fieldKey]);
      } else {
        validationMessages[fieldKey] = {
          errors: [],
          warnings: [],
        };
        validationMessages[fieldKey].errors = [validations[fieldKey]];
      }
    }

    return validationMessages;
  }

  public render(): JSX.Element {
    const { component: { simplified, id } } = this.props;
    const { address, zipCode, postPlace, careOf, houseNumber } = this.state;

    const validations = this.joinValidationMessages();

    if (simplified) {
      return (
        <div className={'address-component'}>
          <label className={'address-component-label'}>
            {
              // This has been implemented for the sake of validating new textResource binding POC
              (!this.props.component.textResourceBindings.address) ? 'Adresse' :
                this.props.getTextResource(this.props.component.textResourceBindings.address)
            }
          </label>
          <input
            className={'form-control' + (this.props.component.readOnly ? ' disabled' : '')}
            value={address}
            onChange={this.updateField.bind(null, AddressKeys.address)}
            onBlur={this.onBlurField.bind(null, AddressKeys.address)}
            disabled={this.props.component.readOnly}
          />
          {validations ?
            renderValidationMessagesForComponent(validations[AddressKeys.address], `${id}_${AddressKeys.address}`)
            : null}
          <div className={'address-component-postplace-zipCode'}>
            <div className={'address-component-zipCode'}>
              <label className={'address-component-label'}>Postnummer</label>
              <input
                className={classNames('address-component-small-inputs', 'form-control',
                  {
                    'validation-error': (validations ? validations.zipCode : false),
                    'disabled': this.props.component.readOnly,
                  })}
                value={zipCode}
                onChange={this.updateField.bind(null, AddressKeys.zipCode)}
                onBlur={this.onBlurField.bind(null, AddressKeys.zipCode)}
                disabled={this.props.component.readOnly}
              />
              {validations ?
                renderValidationMessagesForComponent(validations[AddressKeys.zipCode], `${id}_${AddressKeys.zipCode}`)
                : null}
            </div>

            <div className={'address-component-postplace'}>
              <label className={'address-component-label'}>Poststed</label>
              <input
                className={classNames('form-control', { 'disabled': this.props.component.readOnly })}
                value={postPlace}
                onChange={this.updateField.bind(null, AddressKeys.postPlace)}
                onBlur={this.onBlurField.bind(null, AddressKeys.postPlace)}
                disabled={this.props.component.readOnly}
              />
              {validations ?
                renderValidationMessagesForComponent(validations[AddressKeys.postPlace],
                  `${id}_${AddressKeys.postPlace}`)
                : null}
            </div>
          </div>
        </div>
      );
    }
    return (
      <div className={'address-component'}>
        <label className={'address-component-label'}>{
          // This has been implemented for the sake of validating new textResource binding POC
          (!this.props.component.textResourceBindings.address) ? 'Adresse' :
            this.props.getTextResource(this.props.component.textResourceBindings.address)
        }</label>
        <input
          className={classNames('form-control', { 'disabled': this.props.component.readOnly })}
          value={address}
          onChange={this.updateField.bind(null, AddressKeys.address)}
          onBlur={this.onBlurField.bind(null, AddressKeys.address)}
        />
        {validations ?
          renderValidationMessagesForComponent(validations[AddressKeys.address], `${id}_${AddressKeys.address}`)
          : null}
        <label className={'address-component-label'}>c/o eller annen tilleggsadresse</label>
        <input
          className={classNames('form-control', { 'disabled': this.props.component.readOnly })}
          value={careOf}
          onChange={this.updateField.bind(null, AddressKeys.careOf)}
          onBlur={this.onBlurField.bind(null, AddressKeys.careOf)}
        />
        {validations ?
          renderValidationMessagesForComponent(validations[AddressKeys.careOf], `${id}_${AddressKeys.careOf}`)
          : null}
        <div className={'address-component-postplace-zipCode'}>
          <div className={'address-component-zipCode'}>
            <label className={'address-component-label'}>Postnummer</label>
            <br />
            <input
              className={classNames('address-component-small-inputs', 'form-control',
                {
                  'validation-error': (validations ? validations.zipCode : false),
                  'disabled': this.props.component.readOnly,
                })}
              value={zipCode}
              onChange={this.updateField.bind(null, AddressKeys.zipCode)}
              onBlur={this.onBlurField.bind(null, AddressKeys.zipCode)}
            />
            {validations ?
              renderValidationMessagesForComponent(validations[AddressKeys.zipCode], `${id}_${AddressKeys.zipCode}`)
              : null}
          </div>
          <div className={'address-component-postplace'}>
            <label className={'address-component-label'}>Poststed</label>
            <br />
            <input
              className={classNames('form-control', { 'disabled': this.props.component.readOnly })}
              value={postPlace}
              onChange={this.updateField.bind(null, AddressKeys.postPlace)}
              onBlur={this.onBlurField.bind(null, AddressKeys.postPlace)}
            />
            {validations ?
              renderValidationMessagesForComponent(validations[AddressKeys.postPlace], `${id}_${AddressKeys.postPlace}`)
              : null}
          </div>
        </div>
        <label className={'address-component-label'}>
          Bolignummer
          <label className={'address-component-label-smaller'}>
            <span>&nbsp;</span>(Valgfri)
          </label>
        </label>
        <p>
          Om addressen er felles for flere boenhenter må du oppgi bolignummer.
          Den består av en bokstav og fire tall og skal være ført opp ved/på inngangsdøren din.
        </p>
        <input
          className={classNames('address-component-small-inputs', 'form-control',
            {
              'validation-error': (validations ? validations.houseNumber : false),
              'disabled': this.props.component.readOnly,
            })}
          value={houseNumber}
          onChange={this.updateField.bind(null, AddressKeys.houseNumber)}
          onBlur={this.onBlurField.bind(null, AddressKeys.houseNumber)}
        />
        {validations ?
          renderValidationMessagesForComponent(validations[AddressKeys.houseNumber], `${id}_${AddressKeys.houseNumber}`)
          : null}
      </div>
    );
  }
}

export function getTextResourceByAddressKey(key: AddressKeys, language: any): string {
  switch (key) {
    case AddressKeys.address: {
      return language.ux_editor.modal_configure_address_component_address;
    }
    case AddressKeys.zipCode: {
      return language.ux_editor.modal_configure_address_component_zip_code;
    }
    case AddressKeys.houseNumber: {
      return language.ux_editor.modal_configure_address_component_house_number;
    }
    case AddressKeys.careOf: {
      return language.ux_editor.modal_configure_address_component_care_of;
    }
    case AddressKeys.postPlace: {
      return language.ux_editor.modal_configure_address_component_post_place;
    }
  }
}
