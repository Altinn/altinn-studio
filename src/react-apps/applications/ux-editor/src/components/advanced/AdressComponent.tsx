import update from 'immutability-helper';
import * as React from 'react';

import '../../styles/AdressComponent.css';

export interface IAdressComponentProps {
  component: IFormComponent;
  formData: any;
  handleDataChange: (value: any) => void;
  isValid?: boolean;
  simplified: boolean;
}

export interface IAdressComponentState {
  adress: string;
  postNumber: number;
  postPlace: string;
  careOf: string;
  houseNumber: string;
}

enum AdressKeys {
  adress,
  postNumber,
  postPlace,
  careOf,
  houseNumber,
}

export class AdressComponent extends React.Component<IAdressComponentProps, IAdressComponentState> {
  constructor(_props: IAdressComponentProps) {
    super(_props);
    this.state = {
      adress: '',
      postNumber: null,
      postPlace: '',
      careOf: '',
      houseNumber: '',
    };
  }

  public updateField: (key: AdressKeys, event: any) => void = (key: AdressKeys, event: any): void => {
    const changedFieldValue = event.target.value;
    this.setState((state: IAdressComponentState) => update(state, {
      [AdressKeys[key]]: {
        $set: changedFieldValue,
      },
    }));
  }

  public render(): JSX.Element {
    const { component: { simplified } } = this.props;
    const { adress, postNumber, postPlace, careOf, houseNumber } = this.state;

    if (simplified) {
      return (
        <div className={'adress-component'}>
          <label className={'adress-component-label'}>Adresse</label>
          <input
            className={'form-control'}
            value={adress}
            onChange={this.updateField.bind(null, AdressKeys.adress)}
          />
          <div className={'adress-component-postplace-postnumber'}>
            <div className={'adress-component-postnumber'}>
              <label className={'adress-component-label'}>Postnummer</label>
              <input
                className={'adress-component-small-inputs form-control'}
                value={postNumber}
                onChange={this.updateField.bind(null, AdressKeys.postNumber)}
              />
            </div>
            <div className={'adress-component-postplace'}>
              <label className={'adress-component-label'}>Poststed</label>
              <input
                className={'form-control'}
                value={postPlace}
                onChange={this.updateField.bind(null, AdressKeys.postPlace)}
              />
            </div>
          </div>
        </div>
      );
    }
    return (
      <div className={'adress-component'}>
        <label className={'adress-component-label'}>Adresse</label>
        <input
          className={'form-control'}
          value={adress}
          onChange={this.updateField.bind(null, AdressKeys.adress)}
        />
        <label className={'adress-component-label'}>c/o eller annen tilleggsadresse</label>
        <input
          className={'form-control'}
          value={careOf}
          onChange={this.updateField.bind(null, AdressKeys.careOf)}
        />
        <div className={'adress-component-postplace-postnumber'}>
          <div className={'adress-component-postnumber'}>
            <label className={'adress-component-label'}>Postnummer</label>
            <br />
            <input
              className={'adress-component-small-inputs form-control'}
              value={postNumber}
              onChange={this.updateField.bind(null, AdressKeys.postNumber)}
            />
          </div>
          <div className={'adress-component-postplace'}>
            <label className={'adress-component-label'}>Poststed</label>
            <br />
            <input
              className={'form-control'}
              value={postPlace}
              onChange={this.updateField.bind(null, AdressKeys.postPlace)}
            />
          </div>
        </div>
        <label className={'adress-component-label'}>
          Bolignummer
          <label className={'adress-component-label-smaller'}>
            <span>&nbsp;</span>(Valgfri)
          </label>
        </label>
        <p>
          Om adressen er felles for flere boenhenter må du oppgi bolignummer.
          Den består av en bokstav og fire tall og skal være ført opp ved/på inngangsdøren din.
        </p>
        <input
          className={'adress-component-small-inputs form-control'}
          value={houseNumber}
          onChange={this.updateField.bind(null, AdressKeys.houseNumber)}
        />
      </div>
    );
  }
}
