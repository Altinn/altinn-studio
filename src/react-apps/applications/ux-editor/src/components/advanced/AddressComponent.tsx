import axios, { AxiosResponse } from 'axios';
import update from 'immutability-helper';
import * as React from 'react';

import '../../styles/AddressComponent.css';

export interface IAddressComponentProps {
  component: IFormAddressComponent;
  formData: { [id: string]: string };
  handleDataChange: (value: any, key: string) => void;
  getTextResource: (key: string) => string;
  isValid?: boolean;
  simplified: boolean;
}

interface IAddressValidationErrors {
  zipCode?: string;
  houseNumber?: string;
}

export interface IAddressComponentState {
  address: string;
  zipCode: string;
  postPlace: string;
  careOf: string;
  houseNumber: string;
  validations: IAddressValidationErrors;
}

export enum AddressKeys {
  address = 'address',
  zipCode = 'zipCode',
  postPlace = 'postPlace',
  careOf = 'careOf',
  houseNumber = 'houseNumber',
}

export class AddressComponent extends React.Component<IAddressComponentProps, IAddressComponentState> {
  constructor(_props: IAddressComponentProps) {
    super(_props);
    const formData = this.props.formData ? this.props.formData : {};
    this.state = {
      address: formData.address || '',
      zipCode: formData.zipCode || '',
      postPlace: formData.postPlace || '',
      careOf: formData.careOf || '',
      houseNumber: formData.houseNumber || '',
      validations: {
        zipCode: null,
        houseNumber: null,
      },
    };
  }

  public fetchPostPlace: (zipCode: string) => void = (zipCode: string) => {
    if (zipCode.match(new RegExp('^[0-9]{4}$'))) {
      axios.get('https://api.bring.com/shippingguide/api/postalCode.json', {
        params: {
          clientUrl: window.location.href,
          pnr: zipCode,
        },
      }).then((response: AxiosResponse) => {
        if (response.data.valid) {
          this.setState({
            postPlace: response.data.result,
            validations: {
              zipCode: null,
            },
          }, () => {
            this.onBlurField(AddressKeys.postPlace);
          });
        } else {
          this.setState({
            postPlace: '',
            validations: {
              zipCode: 'Postnummer er ikke gyldig',
            },
          }, () => {
            this.onBlurField(AddressKeys.postPlace);
          });
        }
      });
      this.setState({
        postPlace: 'Laster...',
      });
    }
  }

  public validate: () => IAddressValidationErrors = () => {
    const { zipCode, houseNumber } = this.state;
    const validationErrors: IAddressValidationErrors = {
      zipCode: null,
      houseNumber: null,
    };
    if (!zipCode.match(new RegExp('^[0-9]{4}$')) && zipCode !== '') {
      validationErrors.zipCode = 'Postnummer er ikke gyldig';
    } else {
      validationErrors.zipCode = null;
    }
    if (!houseNumber.match(new RegExp('^[a-z,A-Z]{1}[0-9]{4}$')) && houseNumber !== '') {
      validationErrors.houseNumber = 'Bolignummer er ikke gyldig';
    } else {
      validationErrors.houseNumber = null;
    }
    return validationErrors;
  }

  public updateField: (key: AddressKeys, event: any) => void = (key: AddressKeys, event: any): void => {
    const changedFieldValue: string = event.target.value;
    const changedKey: string = AddressKeys[key];
    this.setState((state: IAddressComponentState) => update(state, {
      [changedKey]: {
        $set: changedFieldValue,
      },
    }));
  }

  public onBlurField: (key: AddressKeys) => void = (key: AddressKeys) => {
    const validationErrors: IAddressValidationErrors = this.validate();
    if (!validationErrors.zipCode && !validationErrors.houseNumber) {
      this.props.handleDataChange(this.state[key], key);
      if (AddressKeys[key] === 'zipCode') {
        this.fetchPostPlace(this.state[key]);
      }
    }
    this.setState({
      validations: validationErrors,
    });
  }

  public render(): JSX.Element {
    const { component: { simplified } } = this.props;
    const { address, zipCode, postPlace, careOf, houseNumber, validations } = this.state;

    if (simplified) {
      return (
        <div className={'address-component'}>
          <label className={'address-component-label'}>
            {
              // This has been implemented for the sake of validating new textResource binding POC
              (!this.props.component.textResourceBindings.address) ? 'Adresse' :
                this.props.getTextResource(this.props.component.textResourceBindings.address)
            }
          </label>
          <input
            className={'form-control'}
            value={address}
            onChange={this.updateField.bind(null, AddressKeys.address)}
            onBlur={this.onBlurField.bind(null, AddressKeys.address)}
          />
          <div className={'address-component-postplace-zipCode'}>
            <div className={'address-component-zipCode'}>
              <label className={'address-component-label'}>Postnummer</label>
              <input
                className={
                  !validations.zipCode ?
                    'address-component-small-inputs form-control' :
                    'address-component-small-inputs form-control validation-error'
                }
                value={zipCode}
                onChange={this.updateField.bind(null, AddressKeys.zipCode)}
                onBlur={this.onBlurField.bind(null, AddressKeys.zipCode)}
              />
            </div>
            <div className={'address-component-postplace'}>
              <label className={'address-component-label'}>Poststed</label>
              <input
                className={'form-control'}
                value={postPlace}
                onChange={this.updateField.bind(null, AddressKeys.postPlace)}
                onBlur={this.onBlurField.bind(null, AddressKeys.postPlace)}
              />
            </div>
          </div>
          {!validations.zipCode ?
            null :
            <label
              className={'address-component-validation-label'}
            >
              {validations.zipCode}
            </label>
          }
        </div>
      );
    }
    return (
      <div className={'address-component'}>
        <label className={'address-component-label'}>{
          // This has been implemented for the sake of validating new textResource binding POC
          (!this.props.component.textResourceBindings.address) ? 'Adresse' :
            this.props.getTextResource(this.props.component.textResourceBindings.address)
        }</label>
        <input
          className={'form-control'}
          value={address}
          onChange={this.updateField.bind(null, AddressKeys.address)}
          onBlur={this.onBlurField.bind(null, AddressKeys.address)}
        />
        <label className={'address-component-label'}>c/o eller annen tilleggsadresse</label>
        <input
          className={'form-control'}
          value={careOf}
          onChange={this.updateField.bind(null, AddressKeys.careOf)}
          onBlur={this.onBlurField.bind(null, AddressKeys.careOf)}
        />
        <div className={'address-component-postplace-zipCode'}>
          <div className={'address-component-zipCode'}>
            <label className={'address-component-label'}>Postnummer</label>
            <br />
            <input
              className={
                !validations.zipCode ?
                  'address-component-small-inputs form-control' :
                  'address-component-small-inputs form-control validation-error'
              }
              value={zipCode}
              onChange={this.updateField.bind(null, AddressKeys.zipCode)}
              onBlur={this.onBlurField.bind(null, AddressKeys.zipCode)}
            />
          </div>
          <div className={'address-component-postplace'}>
            <label className={'address-component-label'}>Poststed</label>
            <br />
            <input
              className={'form-control'}
              value={postPlace}
              onChange={this.updateField.bind(null, AddressKeys.postPlace)}
              onBlur={this.onBlurField.bind(null, AddressKeys.postPlace)}
            />
          </div>
        </div>
        {!validations.zipCode ?
          null :
          <label
            className={'address-component-validation-label'}
          >
            {validations.zipCode}
          </label>
        }
        <label className={'address-component-label'}>
          Bolignummer
          <label className={'address-component-label-smaller'}>
            <span>&nbsp;</span>(Valgfri)
          </label>
        </label>
        <p>
          Om addressen er felles for flere boenhenter må du oppgi bolignummer.
          Den består av en bokstav og fire tall og skal være ført opp ved/på inngangsdøren din.
        </p>
        <input
          className={
            !validations.houseNumber ?
              'address-component-small-inputs form-control' :
              'address-component-small-inputs form-control validation-error'
          }
          value={houseNumber}
          onChange={this.updateField.bind(null, AddressKeys.houseNumber)}
          onBlur={this.onBlurField.bind(null, AddressKeys.houseNumber)}
        />
        {!validations.houseNumber ?
          null :
          <label
            className={'address-component-validation-label'}
          >
            {validations.houseNumber}
          </label>
        }
      </div>
    );
  }
}

export function getTextResourceByAddressKey(key: AddressKeys, language: any): string {
  switch (key) {
    case AddressKeys.address: {
      return language.ux_editor.modal_configure_address_component_address;
    }
    case AddressKeys.zipCode: {
      return language.ux_editor.modal_configure_address_component_zip_code;
    }
    case AddressKeys.houseNumber: {
      return language.ux_editor.modal_configure_address_component_house_number;
    }
    case AddressKeys.careOf: {
      return language.ux_editor.modal_configure_address_component_care_of;
    }
    case AddressKeys.postPlace: {
      return language.ux_editor.modal_configure_address_component_post_place;
    }
  }
}
