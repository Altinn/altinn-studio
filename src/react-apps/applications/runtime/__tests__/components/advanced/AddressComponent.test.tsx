/* tslint:disable:jsx-wrap-multiline */
import { mount, shallow } from 'enzyme';
import 'jest';
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { AddressComponent, getTextResourceByAddressKey } from '../../../src/components/advanced/AddressComponent';

export interface ITextResourceBindings {
  [id: string]: string;
}
export enum AddressKeys {
  address = 'address',
  zipCode = 'zipCode',
  postPlace = 'postPlace',
  careOf = 'careOf',
  houseNumber = 'houseNumber',
}

describe('>>> components/advanced/AddressComponent.tsx snapshot', () => {
  let mockId: string;
  let mockFormData: { [id: string]: string };
  let mockHandleDataChange: (value: any, key: string) => void;
  let mockGetTextResource: (key: string) => string;
  let mockIsValid: boolean;
  let mockSimplified: boolean;
  let mockValidationMessages: any;
  let mockDataBinding: any;
  let mockReadOnly: boolean;

  mockId = 'mock-id';
  mockFormData = { address: 'adresse 1' };
  mockHandleDataChange = (data: any) => null;
  mockGetTextResource = (resourceKey: string) => 'test';
  mockIsValid = true;
  mockSimplified = true;
  mockReadOnly = false;
  mockDataBinding = {};
  mockValidationMessages = {
    zipCode: null,
    houseNumber: null,
  };

  it('>>> Capture snapshot of AddressComponent', () => {
    const rendered = renderer.create(
      <AddressComponent
        id={mockId}
        formData={mockFormData}
        handleDataChange={mockHandleDataChange}
        getTextResource={mockGetTextResource}
        isValid={mockIsValid}
        simplified={mockSimplified}
        dataModelBindings={mockDataBinding}
        validationMessages={mockValidationMessages}
        readOnly={mockReadOnly}
      />,
    );
    expect(rendered).toMatchSnapshot();
  });
  it('+++ should have correct state', () => {
    const shallowAddressComponent = shallow(
      <AddressComponent
        id={mockId}
        formData={mockFormData}
        handleDataChange={mockHandleDataChange}
        getTextResource={mockGetTextResource}
        isValid={mockIsValid}
        simplified={mockSimplified}
        dataModelBindings={mockDataBinding}
        validationMessages={mockValidationMessages}
        readOnly={mockReadOnly}
      />,
    );
    const instance = shallowAddressComponent.instance() as AddressComponent;
    expect(instance.state.address).toEqual('adresse 1');
    expect(instance.state.zipCode).toEqual('');
  });
  it('+++ should return simplified version when simplified', () => {
    const shallowAddressComponent = shallow(
      <AddressComponent
        id={mockId}
        formData={mockFormData}
        handleDataChange={mockHandleDataChange}
        getTextResource={mockGetTextResource}
        isValid={mockIsValid}
        simplified={mockSimplified}
        dataModelBindings={mockDataBinding}
        validationMessages={mockValidationMessages}
        readOnly={mockReadOnly}
      />,
    );
    expect(shallowAddressComponent.find('input').length).toBe(3);
  });
  it('+++ should return advanced version when not simplified', () => {
    const mountedAddressComponent = mount(
      <AddressComponent
        id={mockId}
        formData={mockFormData}
        handleDataChange={mockHandleDataChange}
        getTextResource={mockGetTextResource}
        isValid={mockIsValid}
        simplified={false}
        dataModelBindings={mockDataBinding}
        validationMessages={mockValidationMessages}
        readOnly={mockReadOnly}
      />,
    );
    const instance = mountedAddressComponent.instance() as AddressComponent;
    instance.setState({ ...mockValidationMessages });
    const spy = jest.spyOn(instance, 'joinValidationMessages');
    instance.render();

    expect(spy).toHaveBeenCalled();
    expect(mountedAddressComponent.find('input').length).toBe(5);
  });
  it('+++ should render editable component when readOnly is false', () => {
    const shallowAddressComponent = shallow(
      <AddressComponent
        id={mockId}
        formData={mockFormData}
        handleDataChange={mockHandleDataChange}
        getTextResource={mockGetTextResource}
        isValid={mockIsValid}
        simplified={mockSimplified}
        dataModelBindings={mockDataBinding}
        validationMessages={mockValidationMessages}
        readOnly={mockReadOnly}
      />,
    );
    expect(shallowAddressComponent.find('.address-component-small-inputs').hasClass('disabled')).toBe(false);
  });
  it('+++ should render disabled component when readOnly is true', () => {
    const shallowAddressComponent = shallow(
      <AddressComponent
        id={mockId}
        formData={mockFormData}
        handleDataChange={mockHandleDataChange}
        getTextResource={mockGetTextResource}
        isValid={mockIsValid}
        simplified={mockSimplified}
        dataModelBindings={mockDataBinding}
        validationMessages={mockValidationMessages}
        readOnly={true}
      />,
    );
    shallowAddressComponent.find('input').forEach((node: any) => {
      expect(node.hasClass('disabled')).toBe(true);
    });
  });
  it('+++ should return advanced version and disabled when not simplified and readOnly is true', () => {
    const shallowAddressComponent = shallow(
      <AddressComponent
        id={mockId}
        formData={mockFormData}
        handleDataChange={mockHandleDataChange}
        getTextResource={mockGetTextResource}
        isValid={mockIsValid}
        simplified={false}
        dataModelBindings={mockDataBinding}
        validationMessages={mockValidationMessages}
        readOnly={true}
      />,
    );
    shallowAddressComponent.find('input').forEach((node: any) => {
      expect(node.hasClass('disabled')).toBe(true);
    });
    expect(shallowAddressComponent.find('input').length).toBe(5);
  });
  it('+++ should run onBlurField on blur + fetch Post place zipCode blur', () => {
    const mountedAddressComponent = mount(
      <AddressComponent
        id={mockId}
        formData={mockFormData}
        handleDataChange={mockHandleDataChange}
        getTextResource={mockGetTextResource}
        isValid={mockIsValid}
        simplified={mockSimplified}
        dataModelBindings={mockDataBinding}
        validationMessages={mockValidationMessages}
        readOnly={mockReadOnly}
      />,
    );
    const instance = mountedAddressComponent.instance() as AddressComponent;
    const spyOnBlurField = jest.spyOn(instance, 'onBlurField');
    const spyFetchPostPlace = jest.spyOn(instance, 'fetchPostPlace');
    instance.forceUpdate();

    const inputFields = mountedAddressComponent.find('input');
    inputFields.forEach((input) => {
      input.simulate('blur');
      expect(spyOnBlurField).toHaveBeenCalled();
    });
    mountedAddressComponent.find('.address-component-zipCode').find('input').simulate('blur');
    expect(spyFetchPostPlace).toHaveBeenCalled();
    expect(instance.state.postPlace).toEqual('');
  });
  it('+++ should trigger updateField onChange', () => {
    const mountedAddressComponent = mount(
      <AddressComponent
        id={mockId}
        formData={mockFormData}
        handleDataChange={mockHandleDataChange}
        getTextResource={mockGetTextResource}
        isValid={mockIsValid}
        simplified={mockSimplified}
        dataModelBindings={mockDataBinding}
        validationMessages={mockValidationMessages}
        readOnly={mockReadOnly}
      />,
    );
    const instance = mountedAddressComponent.instance() as AddressComponent;
    const spyUpdateField = jest.spyOn(instance, 'updateField');
    instance.forceUpdate();

    const inputField = mountedAddressComponent.find('input').first();
    inputField.simulate('change');
    expect(spyUpdateField).toHaveBeenCalled();
  });
  it('+++ getTextResourceByAddressKey should return the correct key', () => {
    let mockKey = AddressKeys.address;
    const mockLanguage = {
      ux_editor: {
        modal_configure_address_component_address: 'Adresse',
        modal_configure_address_component_zip_code: 'Postnr',
        modal_configure_address_component_house_number: 'Bolignummer',
        modal_configure_address_component_care_of: 'C/O eller annen tilleggsadresse',
        modal_configure_address_component_post_place: 'Poststed',
      },
    };
    const result1 = getTextResourceByAddressKey(mockKey, mockLanguage);
    expect(result1).toEqual('Adresse');

    mockKey = AddressKeys.zipCode;
    const result2 = getTextResourceByAddressKey(mockKey, mockLanguage);
    expect(result2).toEqual('Postnr');

    mockKey = AddressKeys.houseNumber;
    const result3 = getTextResourceByAddressKey(mockKey, mockLanguage);
    expect(result3).toEqual('Bolignummer');

    mockKey = AddressKeys.careOf;
    const result4 = getTextResourceByAddressKey(mockKey, mockLanguage);
    expect(result4).toEqual('C/O eller annen tilleggsadresse');

    mockKey = AddressKeys.postPlace;
    const result5 = getTextResourceByAddressKey(mockKey, mockLanguage);
    expect(result5).toEqual('Poststed');
  });

  it('+++ joinValidationMessages should return all validations in correct format', () => {
    const mockResult = {
      address: {
        errors: [],
        warnings: [],
      },
      careOf: {
        errors: [],
        warnings: [],
      },
      houseNumber: {
        errors: ['some other error'],
        warnings: [],
      },
      postPlace: {
        errors: [],
        warnings: [],
      },
      zipCode: {
        errors: ['some error'],
        warnings: [],
      },
    };
    const mountedAddressComponent = mount(
      <AddressComponent
        id={mockId}
        formData={mockFormData}
        handleDataChange={mockHandleDataChange}
        getTextResource={mockGetTextResource}
        isValid={mockIsValid}
        simplified={mockSimplified}
        dataModelBindings={mockDataBinding}
        readOnly={mockReadOnly}
      />,
    );
    const instance = mountedAddressComponent.instance() as AddressComponent;
    instance.setState({
      validations: {
        zipCode: 'some error',
        houseNumber: 'some other error',
      },
    });
    const result = instance.joinValidationMessages();
    expect(result).toEqual(mockResult);
  });

});
