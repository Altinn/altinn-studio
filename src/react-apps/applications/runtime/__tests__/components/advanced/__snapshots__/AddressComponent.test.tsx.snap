// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> components/advanced/AddressComponent.tsx snapshot >>> Capture snapshot of AddressComponent 1`] = `
<div
  className="address-component"
>
  <label
    className="address-component-label"
  >
    Adresse
  </label>
  <input
    className="form-control"
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    value=""
  />
  <div
    className="address-component-postplace-zipCode"
  >
    <div
      className="address-component-zipCode"
    >
      <label
        className="address-component-label"
      >
        Postnummer
      </label>
      <input
        className="address-component-small-inputs form-control"
        disabled={false}
        onBlur={[Function]}
        onChange={[Function]}
        value=""
      />
    </div>
    <div
      className="address-component-postplace"
    >
      <label
        className="address-component-label"
      >
        Poststed
      </label>
      <input
        className="form-control"
        disabled={false}
        onBlur={[Function]}
        onChange={[Function]}
        value=""
      />
    </div>
  </div>
</div>
`;
